### PR TITLE
Add submission cancel feature

### PR DESCRIPTION
**Pull request by Ram81**
### Description

This PR adds - 

- [x] Allow user to cancel submissions if the status is `SUBMITTED`

### CODE REVIEW

File: apps/jobs/urls.py [hunk 1]
```diff
@@ -110,6 +110,11 @@
         views.update_submission_started_at,
         name="update_submission_started_at",
     ),
+    url(
+        r"^challenge/(?P<challenge_pk>[0-9]+)/submissions/(?P<submission_pk>[0-9]+)/cancel/$",
```
Comment (RishabhJain2018) [inline apps/jobs/urls.py:114]: Why do we need `challenge_pk` here? I think `submission_pk` should be enough here, no?

File: apps/jobs/urls.py [hunk 2]
```diff
@@ -110,6 +110,11 @@
...
+        views.change_submission_status_to_cancel,
```
Comment (RishabhJain2018) [inline apps/jobs/urls.py:115]: We should create a generic API that changes the status of the submissions to other status as well.

File: apps/jobs/views.py [hunk 3]
```diff
@@ -2775,3 +2775,46 @@ def update_submission_started_at(request, submission_pk):
         return Response(response_data, status=status.HTTP_200_OK)
     else:
         return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+
+@api_view(["PATCH"])
+@throttle_classes([UserRateThrottle])
+@permission_classes((permissions.IsAuthenticated, HasVerifiedEmail))
+@authentication_classes((JWTAuthentication, ExpiringTokenAuthentication))
+def change_submission_status_to_cancel(request, challenge_pk, submission_pk):
```
Comment (RishabhJain2018) [inline apps/jobs/views.py:2784]: I think `submission_pk` should be enough here?

File: apps/jobs/views.py [hunk 4]
```diff
@@ -2775,3 +2775,46 @@ def update_submission_started_at(request, submission_pk):
...
+    """
+    API Endpoint for setting submission status to CANCELLED.
+    """
+    print(request.user, challenge_pk)
```
Comment (RishabhJain2018) [inline apps/jobs/views.py:2788]: Please remove the print statement here.

File: apps/jobs/views.py [hunk 5]
```diff
@@ -2775,3 +2775,46 @@ def update_submission_started_at(request, submission_pk):
...
+
+    try:
+        participant_team = ParticipantTeam.objects.get(pk=participant_team_pk)
+    except ParticipantTeam.DoesNotExist:
+        response_data = {"error": "You haven't participated in the challenge"}
+        return Response(response_data, status=status.HTTP_403_FORBIDDEN)
+
+    try:
+        submission = Submission.objects.get(
+            id=submission_pk,
```
Comment (RishabhJain2018) [inline apps/jobs/views.py:2801]: We can simply get the Submission from the `submission_pk`, why do we need participant team here?

File: frontend/src/js/controllers/challengeCtrl.js [hunk 6]
```diff
@@ -1991,7 +1991,42 @@
             }
         };
 
-        vm.hideVisibilityDialog = function() {
+        vm.cancelSubmission = function(submissionId) {
+            parameters.url = "jobs/challenge/" + vm.challengeId + "/submissions/" + submissionId + "/cancel/";
+            parameters.method = 'PATCH';
+            console.log("Cancelling submission: " + vm.submissionId);
```
Comment (RishabhJain2018) [inline frontend/src/js/controllers/challengeCtrl.js:1997]: Please remove console statement.

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]
Can you please also add the screenshot of the UI?

File: apps/jobs/urls.py (hunk 1 above)
Comment (Ram81) [reply apps/jobs/urls.py:114]: We only want the participants of the team to cancel the submission using UI. To check whether team has participated in the challenge we need challenge pk

File: apps/jobs/urls.py (hunk 2 above)
Comment (Ram81) [reply apps/jobs/urls.py:115]: This API is specific to the user functionality of cancelling submission. Why should this one be allowed to set other statuses?

File: apps/jobs/views.py (hunk 3 above)
Comment (Ram81) [reply apps/jobs/views.py:2784]: We only want the participants of the team to cancel the submission using UI. To check whether team has participated in the challenge we need challenge pk

File: apps/jobs/views.py (hunk 5 above)
Comment (Ram81) [reply apps/jobs/views.py:2801]: Addressed in above comments

File: apps/jobs/views.py (hunk 4 above)
Comment (RishabhJain2018) [reply apps/jobs/views.py:2848]: @Ram81 Can you please remove it?

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]

File: apps/jobs/views.py [hunk 7]
```diff
@@ -2807,7 +2807,74 @@ def update_submission_meta(request, challenge_pk, submission_pk):
...
     else:
-        response_data = {
-            "error": "Sorry, you are not authorized to make this request"
-        }
+        participant_team_pk = get_participant_team_id_of_user_for_a_challenge(
+            request.user, challenge_pk
+        )
+
+        try:
+            participant_team = ParticipantTeam.objects.get(pk=participant_team_pk)
```
Comment (RishabhJain2018) [inline apps/jobs/views.py:2815]: Can we please use/create `get_participant_model` like we have for challenges/challenge phases.

File: apps/jobs/views.py [hunk 8]
```diff
@@ -2807,7 +2807,74 @@ def update_submission_meta(request, challenge_pk, submission_pk):
...
+            response_data = {"error": "You haven't participated in the challenge"}
+            return Response(response_data, status=status.HTTP_403_FORBIDDEN)
+
+        try:
+            submission = Submission.objects.get(
+                id=submission_pk,
+                participant_team=participant_team,
+            )
+        except Submission.DoesNotExist:
+            response_data = {"error": "Submission does not exist"}
```
Comment (RishabhJain2018) [inline apps/jobs/views.py:2826]: `{"error": "Submission {} does not exist".format(submission_pk)}`

File: apps/jobs/views.py [hunk 9]
```diff
@@ -2807,7 +2807,74 @@ def update_submission_meta(request, challenge_pk, submission_pk):
...
+            return Response(response_data, status=status.HTTP_403_FORBIDDEN)
```
Comment (RishabhJain2018) [inline apps/jobs/views.py:2827]: It should be 404 , no?

File: apps/jobs/views.py (hunk 4 above)
Comment (RishabhJain2018) [reply apps/jobs/views.py:2848]: @Ram81 ?

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]

File: apps/jobs/views.py (hunk 8 above)
Comment (RishabhJain2018) [reply apps/jobs/views.py:2840]: @Ram81

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]

**Review by RishabhJain2018 — APPROVED** [review]

**Pull request merged**